### PR TITLE
Separate calculation logic

### LINKED ISSUE #74: Move all calculation logic to a separate module
As it was written here https://github.com/IraSoro/peri/issues/63#issuecomment-1613810939, we run into the limitation that all necessary calculations are carried out to react hooks, due to which ordinary arithmetic unit tests turn into UI tests

This can be fixed by moving all the calculations to a separate module file, and leaving the hooks as a wrapper for react

### PR DESCRIPTION
**Pull request by IraSoro**
Closed #74 #80

### CODE REVIEW

File: src/state/CalculationLogics.ts [hunk 1]
```diff
@@ -0,0 +1,248 @@
...
+export function getDaysBeforePeriod(cycleLength: number, startDate: string) {
+  if (!startDate || !cycleLength) {
+    return {
+      title: i18n.t("Period in"),
+      days: i18n.t("no info"),
+    };
+  }
+
+  const dateOfFinish = new Date(startDate);
+  dateOfFinish.setDate(dateOfFinish.getDate() + Number(cycleLength));
```
Comment (imblowfish) [inline src/state/CalculationLogics.ts:52]: `cycleLengh` is already a `Number`

**Review by imblowfish — CHANGES_REQUESTED** [review]
I don't agree with how `i18next` used in this module. For example, using the key `Days less 5` and some custom calculations `(dayBefore > 20 && dayBefore % 10 > 0 && dayBefore % 10 < 5)` specific to a particular language instead of plurals interval https://www.i18next.com/translation-function/plurals#interval-plurals

I also see a problem in the current algorithm for calculating the average value in the `getAverageLengthOfCycle` and `getAverageLengthOfPeriod` functions, which are too complicated due to the current approach to storing cycles

if you think you can fix it fast now, then ok, or you can create tasks for later to fix it and fix only my simple comments above

File: src/pages/TabHome.tsx [hunk 2]
```diff
@@ -30,123 +30,86 @@ import MarkModal from "../modals/MarkModal";
...
+  getPregnancyChance,
+  getDaysBeforePeriod,
+} from "../state/CalculationLogics";
+
+const p_style = {
+  fontSize: "10px" as const,
+  color: "var(--ion-color-light)" as const,
+};
+
+const h_style = {
```
Comment (imblowfish) [inline src/pages/TabHome.tsx:44]: Please rename to camelCase

File: src/state/CalculationLogics.ts [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+  if (!cycleLength || !dayOfCycle) {
+    return "";
+  }
+
+  const lutealPhaseLength = 14;
+  const ovulationDay = cycleLength - lutealPhaseLength;
+  const diffDay = ovulationDay - dayOfCycle;
+  if (diffDay === 0) {
+    return i18n.t("today");
+  } else if (diffDay < 0 && diffDay >= -2) {
```
Comment (imblowfish) [inline src/state/CalculationLogics.ts:16]:
```js
if (diffDay < -2) {
  // finished
}
if (diffDay < 0) {
  // possible
} 
if (diffDay === 0){
  // today
}
if (diffDay === 1) {
  // tomorrow
} 
```

File: src/state/CalculationLogics.ts [hunk 4]
```diff
@@ -0,0 +1,219 @@
...
+    return i18n.t("possible");
+  } else if (diffDay < 0) {
+    return i18n.t("finished");
+  } else if (diffDay === 1) {
+    return i18n.t("tomorrow");
+  } else if (diffDay < 5) {
+    return `${i18n.t("in")} ${diffDay} ${i18n.t("Days less 5")}`;
+  }
+  return `${i18n.t("in")} ${diffDay} ${i18n.t("Days")}`;
```
Comment (imblowfish) [inline src/state/CalculationLogics.ts:25]: Forgot about days_interval

File: src/state/CalculationLogics.ts [hunk 5]
```diff
@@ -0,0 +1,219 @@
...
+  if (diffDay === 0) {
+    return i18n.t("today");
+  } else if (diffDay < 0 && diffDay >= -2) {
+    return i18n.t("possible");
+  } else if (diffDay < 0) {
+    return i18n.t("finished");
+  } else if (diffDay === 1) {
+    return i18n.t("tomorrow");
+  } else if (diffDay < 5) {
+    return `${i18n.t("in")} ${diffDay} ${i18n.t("Days less 5")}`;
```
Comment (imblowfish) [inline src/state/CalculationLogics.ts:23]: With days_interval you can delete this expression

File: src/state/CalculationLogics.ts [hunk 6]
```diff
@@ -0,0 +1,219 @@
...
+export function getPregnancyChance(cycleLength: number, dayOfCycle: number) {
+  if (!cycleLength || !dayOfCycle) {
+    return "";
+  }
+
+  const lutealPhaseLength = 14;
+  const ovulationDay = cycleLength - lutealPhaseLength;
+  const diffDay = ovulationDay - dayOfCycle;
+
+  if (diffDay <= 4 && diffDay >= -2) {
```
Comment (imblowfish) [inline src/state/CalculationLogics.ts:37]: Move - 2 expression to the left

File: src/state/CalculationLogics.ts [hunk 7]
```diff
@@ -0,0 +1,219 @@
...
+  const now = new Date();
+  now.setHours(0, 0, 0, 0);
+  const dayBefore = Math.round(
+    (Number(dateOfFinish) - Number(now)) / millisecondsInDay,
+  );
+
+  if (dayBefore > 0) {
+    return {
+      title: i18n.t("Period in"),
+      days: `${dayBefore} ${i18n.t("Days_interval", {
```
Comment (imblowfish) [inline src/state/CalculationLogics.ts:63]: Why here without Math.abs?

File: src/state/CalculationLogics.ts [hunk 8]
```diff
@@ -0,0 +1,219 @@
...
+    return {
+      title: i18n.t("Period"),
+      days: i18n.t("today"),
+    };
+  }
+  return {
+    title: i18n.t("Delay"),
+    days: `${Math.abs(dayBefore)} ${i18n.t("Days_interval", {
+      postProcess: "interval",
+      count: Math.abs(dayBefore),
```
Comment (imblowfish) [inline src/state/CalculationLogics.ts:79]: Double Math.abs calculation

File: src/state/CycleInformationHooks.tsx [hunk 9]
```diff
@@ -1,7 +1,11 @@
 import { useContext } from "react";
 import { CyclesContext } from "./Context";
-
-const millisecondsInDay = 24 * 60 * 60 * 1000;
+import {
+  getDayOfCycle,
+  // getAverageLengthOfCycle,
```
Comment (imblowfish) [inline src/state/CycleInformationHooks.tsx:5]: Forgot comments

File: src/state/CycleInformationHooks.tsx [hunk 10]
```diff
@@ -15,63 +19,17 @@ export function useLastStartDate(): string {
 
 export function useDayOfCycle(): string {
```
Comment (imblowfish) [inline src/state/CycleInformationHooks.tsx:20]: This function contains two lines only, you can remove it

File: src/state/CycleInformationHooks.tsx [hunk 11]
```diff
@@ -1,7 +1,11 @@
...
-
-const millisecondsInDay = 24 * 60 * 60 * 1000;
+import {
+  getDayOfCycle,
+  // getAverageLengthOfCycle,
+  // getAverageLengthOfPeriod,
+  getAverageLength,
+} from "./CalculationLogics";
 
 export function useLastStartDate(): string {
```
Comment (imblowfish) [inline src/state/CycleInformationHooks.tsx:10]: Simple function, you can remove it

File: src/tests/CalculationLogic.test.ts [hunk 12]
```diff
@@ -0,0 +1,347 @@
...
+  getDaysBeforePeriod,
+  getPhase,
+} from "../state/CalculationLogics";
+
+describe("getOvulationStatus", () => {
+  test("cycleLength is 0 and dayOfCycle is 0", () => {
+    expect(getOvulationStatus(0, 0)).toEqual("");
+  });
+
+  test("returns `in 9 Days` if ovulation in 9 Days", () => {
```
Comment (imblowfish) [inline src/tests/CalculationLogic.test.ts:17]: You didnt fix tests' names

**Review by imblowfish — CHANGES_REQUESTED** [review]

File: src/pages/TabDetails.tsx [hunk 13]
```diff
@@ -68,16 +71,10 @@ export function useInfoForOneCycle(idx: number): InfoOneCycle {
     };
   }
   const cycleLenNumber: number = cycles[idx].cycleLength;
```
[the remark this answers is not in the document]
Comment (IraSoro) [reply src/pages/TabDetails.tsx:73]: Fixed

File: src/pages/TabDetails.tsx [hunk 14]
```diff
@@ -33,7 +33,10 @@ function useTitleLastCycle() {
     return "";
   }
 
-  return `${dayOfCycle} ${t("Day")}`;
+  return `${dayOfCycle} ${t("Days_interval", {
+    postProcess: "interval",
+    count: 1,
```
[the remark this answers is not in the document]
Comment (IraSoro) [reply src/pages/TabDetails.tsx:38]: В count должно быть 1, потому что на русском языке в этом моменте пишется, например "28 День", "5 День" и т. д.. В английском языке для обозначения того, какой сейчас день, "День" также пишется в единственном числе, т. е. "Day". Поэтому "count: 1" для того, чтобы написать "День" в единственном числе.

File: src/pages/TabHome.tsx [hunk 15]
```diff
@@ -30,123 +30,86 @@ import MarkModal from "../modals/MarkModal";
...
 import CalendarModal from "../modals/CalendarModal";
 
-const millisecondsInDay = 24 * 60 * 60 * 1000;
+import {
+  getOvulationStatus,
+  getPregnancyChance,
+  getDaysBeforePeriod,
+} from "../state/CalculationLogics";
+
+const p_style = {
```
[the remark this answers is not in the document]
Comment (IraSoro) [reply src/pages/TabHome.tsx:39]: Fixed

File: src/pages/TabHome.tsx (hunk 2 above)
Comment (IraSoro) [reply src/pages/TabHome.tsx:44]: Fixed

File: src/state/CalculationLogics.ts (hunk 3 above)
Comment (IraSoro) [reply src/state/CalculationLogics.ts:16]: Fixed

File: src/state/CalculationLogics.ts (hunk 4 above)
Comment (IraSoro) [reply src/state/CalculationLogics.ts:25]: Fixed

File: src/state/CalculationLogics.ts (hunk 5 above)
Comment (IraSoro) [reply src/state/CalculationLogics.ts:23]: Fixed

File: src/state/CalculationLogics.ts (hunk 6 above)
Comment (IraSoro) [reply src/state/CalculationLogics.ts:37]: Fixed

File: src/state/CalculationLogics.ts (hunk 7 above)
Comment (IraSoro) [reply src/state/CalculationLogics.ts:63]: In this situation, ```dayBefore``` is a positive number. Math.abs is not needed here.

File: src/state/CalculationLogics.ts (hunk 8 above)
Comment (IraSoro) [reply src/state/CalculationLogics.ts:79]: In this situation, ```dayBefore``` is a negative number. (Situation when result is Delay) Math.abs is needed here.

File: src/state/CycleInformationHooks.tsx (hunk 9 above)
Comment (IraSoro) [reply src/state/CycleInformationHooks.tsx:5]: Fixed

File: src/state/CycleInformationHooks.tsx (hunk 11 above)
Comment (IraSoro) [reply src/state/CycleInformationHooks.tsx:10]: Used 2 times

File: src/state/CycleInformationHooks.tsx [hunk 16]
```diff
@@ -15,63 +19,17 @@ export function useLastStartDate(): string {
...
-
-  const diff =
-    Math.ceil((currentDate.getTime() - start.getTime()) / millisecondsInDay) +
-    1;
-
-  return diff.toString();
+  return getDayOfCycle(startDate);
 }
 
 export function useAverageLengthOfCycle(): number {
```
[the remark this answers is not in the document]
Comment (IraSoro) [reply src/state/CycleInformationHooks.tsx:25]: Used 5 times

File: src/state/CycleInformationHooks.tsx [hunk 17]
```diff
@@ -15,63 +19,17 @@ export function useLastStartDate(): string {
...
-      return prev + current.cycleLength;
-    }
-    return 0;
-  }, 0);
-
-  return Math.round(sum / (cycles.length - 1));
+  return getAverageLength(cycles, "cycleLength");
 }
 
 export function useAverageLengthOfPeriod(): number {
```
[the remark this answers is not in the document]
Comment (IraSoro) [reply src/state/CycleInformationHooks.tsx:30]: Fixed

File: src/translations/en.tsx [hunk 18]
```diff
@@ -8,9 +8,9 @@ const en = {
   Period: "Period",
   today: "today",
   Delay: "Delay",
-  Day: "Day",
-  Days: "Days",
-  "Days less 5": "Days", // for russian language
+  Days_one: "Day",
```
[the remark this answers is not in the document]
Comment (IraSoro) [reply src/translations/en.tsx:11]: Fixed

File: src/translations/en.tsx [hunk 19]
```diff
@@ -8,9 +8,9 @@ const en = {
...
+  Days_other: "Days",
```
[the remark this answers is not in the document]
Comment (IraSoro) [reply src/translations/en.tsx:12]: And fixed

File: src/tests/CalculationLogic.test.ts (hunk 12 above)
Comment (IraSoro) [reply src/tests/CalculationLogic.test.ts:17]: Fixed

File: src/pages/TabDetails.tsx (hunk 14 above)
[the remark this answers is not in the document]
Comment (imblowfish) [reply src/pages/TabDetails.tsx:40]: Add NOTE comment

[the remark this answers is not in the document]
Comment (IraSoro) [reply src/pages/TabDetails.tsx:40]: Ok

**Review by imblowfish — APPROVED** [review]

**Pull request merged**